cvlr-spec: basic primitives for specifications

1arie1 opened the pull request:
Introduces basic primitives to write function specification (pre- and post-conditions), invariants, and general Boolean expressions. Expressions can be evaluated, assumed, or asserted. Introduces macros to automatically generate rules from specifications.

1arie1 committed:
cvlr-lock

1arie1 committed:
chore: tests for cvlr_log macro

1arie1 committed:
feat: first draft of cvlr-spec

1arie1 committed:
refactor: split into multiple files

1arie1 committed:
feat: CvlrLemma and related macros

also includes various cleanups in cvlr-spec
and exposure of cvlr-spec in cvlr.

1arie1 committed:
chore: tests and formatting

1arie1 committed:
chore: documentation for CvlrLemma and relations

1arie1 committed:
feat: cvlr_predicate macro to define predicates

Converts a rust function int a CvlrBoolExpr predicate

1arie1 committed:
fix: use __cvlr_ prefix for temporary variables

cvlr_assert series of macros introduce local variables to avoid
evaulating their arguments multiple times. The variable names
should be unique to avoid shadowing.

We are using `__cvlr_` prefix since it is unique enough.

1arie1 committed:
cargo-fmt

1arie1 committed:
fix: use cvlr::spec in generated code

We have cvlr_spec crate that is re-exported in cvlr as cvlr::spec
We expect most users to use cvlr rather than sub-crates individually.

1arie1 committed:
chore: re-export cvlr_predicate in cvlr

re-exporting as cvlr::predicate

1arie1 committed:
feat: adapter to convert one-state to two-state

Useful for converting state predicates to post-conditions

1arie1 committed:
chore: tests and docs for ToTwoState

1arie1 committed:
feat: rewrite cvlr-spec with assoc types

An alternative design in which associated types are used to represent
that context over which a BoolExpr operates.

This design allow programatic access to the context form the BoolExpr.

Also replace StatePair with a tuple of contexts.

1arie1 committed:
chore: tests and docs for new cvlr-spec design

1arie1 committed:
feat: extend BoolExpr to evaluate over states

Currently, only evaluation over two states (or one state) is supported.

Two-state evaluation is used in specification to evaluate over pre- and
post-state, together.

A predicate that is single state is evaluated on the first state of
the two-state pair. For that reason, in CvlrSpec, post-state is given
first.

1arie1 committed:
chore: update cvlr::predicate to match cvlr-spec

1arie1 committed:
chore: update expanded test files

1arie1 committed:
refactor: fmt and clippy warnings

1arie1 committed:
feat: macros to generate rules for specs

1arie1 committed:
feat: macro that generates a single rule

1arie1 committed:
chore: tests for cvlr_rule_for_spec macro

1arie1 committed:
chore: docs and tests for macros in cvlr-spec

1arie1 committed:
chore: cargo-fmt

1arie1 committed:
chore: fix clippy warnings

1arie1 committed:
chore: document methods on CvlrBoolExpr

1arie1 committed:
refactor: rename cvlr_impl to cvlr_implies

1arie1 committed:
chore: clarify that invariant is one-state

1arie1 committed:
chore: cvlr_impl -> cvlr_implies

1arie1 committed:
feat: macros for cvlr_and and cvlr_implies

The macros improve ergonomics of the combinators

1arie1 committed:
chore: documentation and tests for cvlr_and macros

1arie1 committed:
feat: cvlr_pif macro to use pred defining funs

Macro is exposed as `cvlr::pif!(name)`. If `name` is a functions
that defines a `cvlr::predicate`, then pif expands to the predicate
that was defined.

1arie1 committed:
fix: doc tests for cvlr_pif macro

1arie1 committed:
feat: remove guarded expressions from assert_that

1arie1 committed:
feat: assert_that descends into if-else expr

```
cvlr_assert_that!{if guard { flag } else { true }}
```

expands to
```
if guard {
    ::cvlr_asserts::cvlr_assert_checked(flag);
} else {
    ()
};
```

and `cvlr_assert_that!(true)` expands to `()`

1arie1 committed:
chore: in cvlr_predicate, use eager evaluation

Generated `eval` function exits as soon as one of the expressions
is false

caballa commented on cvlr-macros/src/assert_that.rs:
```diff
@@ -88,41 +56,120 @@ fn analyze_condition(condition: &Expr, guard: Option<&Expr>) -> syn::Result<Toke
+    macro_name: &str,
+) -> syn::Result<TokenStream2> {
+    // Check if condition is literal `true`
+    if let Expr::Lit(lit) = condition {
```
How is handled when `condition` is the literal false ?

caballa commented on cvlr-macros/src/lib.rs:
```diff
@@ -431,3 +429,74 @@ pub fn cvlr_eval_that(input: TokenStream) -> TokenStream {
+/// }
+///
+/// // Expands to:
+/// // cvlr_impl_rule!{solvency_update_exchange_price_no_interest_free_new, expr, base_update_exchange_price_no_interest_free_new}
```
it should be `spec` rather than `expr`

caballa commented on cvlr-macros/tests/expand/test_cvlr_predicate.rs:
```diff
@@ -0,0 +1,69 @@
+    assert!(pred5.eval(&ctx));
+    
+    let pred6 = LetBeforeExpressions;
+    assert!(pred6.eval(&ctx));
```
Shouldn't be evaluated to false? `ctx.x` is 5 and inside `LetBeforeExpressions` the first condition is `c.x > threshold` but threshold is 5

caballa commented on cvlr-spec/src/lib.rs:
```diff
@@ -0,0 +1,71 @@
+//! ## Specifications
+//!
+//! The [`CvlrSpec`] trait represents a complete specification with:
+//! - Preconditions (requires) - conditions that must hold before an operation
```
minor thing but maybe change the description of requires to "conditions that are assumed before an operation".
Having the same description for requires and ensures is weird since they do different things

1arie1 replied:
fixed

caballa reviewed (approved):
Amazing

1arie1 committed:
feat: log scope start and end in asserts/assumes

1arie1 committed:
chore: allow trailing semicolon in cvlr_spec

1arie1 committed:
chore: fix tests for assert/assume macros

1arie1 committed:
chore: rename CvlrBoolExpr into CvlrFormula

1arie1 committed:
feat: introduce CvlrPredicate trait

CvlrPredicate is a marker trait that all predicates must implement.

A CvlrFormula is either a constnat CvlrTrue, a predicate, or a boolean
combination of CvlrFormulas.

1arie1 committed:
chore: more tests for cvlr_predicate

The case of if-without-else is broken for evaluation.
The current eager evaluation scheme is difficult to implement in this
case.

Let binidngs are also broken because they are collected now separately
from the expression statements.

Fixing this requires a more serious overhaul.

1arie1 committed:
chore: minor issues in documentation

1arie1 committed:
fix: switch cvlr_predicate to use lazy evaluation

This is on the way of supporting if-without-else, but in the current
variant it still does not work as expected.

1arie1 committed:
chore: runtime test for cvlr_predicate macro

1arie1 committed:
chore: reimplement cvlr_predicate macro

Using cvlr::predicate attributed local functions instead of
cvlr_def_predicate macro.

1arie1 committed:
feat: add form 2 of cvlr_lemma macro

Add form 2 of cvlr_lemma macro that allows for pre-built predicates.

This form is useful when you already have predicates defined or want to use
composed expressions.

update tests and documentation to match

1arie1 committed:
chore: cargo-lock

1arie1 committed:
chore: fix dependencies for cvlr-log tests

1arie1 committed:
chore: cargo-lock

1arie1 committed:
chore: better syntax for cvlr_lemma! macro

1arie1 committed:
chore: reanme bool_expr.rs to formula.rs

1arie1 committed:
fix: close scope in assert macros

assert macros log assertion and values. These are enclosed in a scope.
The bug was not closing the scope correctly using log_scope_end.

1arie1 committed:
chore: make Lemmas public

Lemmas are intendend to be used in multiple contexts.
Make them public by default.

We might add explicit visibility modifier later if needed.

1arie1 committed:
feat: two-state predicate in cvlr_predicate macro

Add support for two-state predicates in cvlr_predicate macro.

Two-state predicates are predicates that take two arguments:
- The first argument is the post-state context
- The second argument is the pre-state context

The macro generates a struct that implements CvlrFormula and
CvlrPredicate.

1arie1 committed:
chore: update expand tests after all the changes

1arie1 committed:
chore: fmt on cvlr::predicate macro

1arie1 committed:
chore: fmt on tests

1arie1 committed:
fix: incorrect order of assume and assert in lemma

CvlrLemma::apply() was essentially same as verify()

1arie1 committed:
feat: cvlr_and_pif macro

Add cvlr_and_pif macro to cvlr-spec.

cvlr_and_pif is a macro that composes a list of predicates
using cvlr_and.
